src/crypto/crypto_context.cc

```diff
@@ -222,6 +226,232 @@ unsigned long LoadCertsFromFile(  // NOLINT(runtime/int)
+      nullptr, 1, &kCFTypeArrayCallBacks);
+  CFArraySetValueAtIndex(subjCerts, 0, ref);
+
+  SecPolicyRef policy = SecPolicyCreateBasicX509();
+  OSStatus ortn = SecTrustCreateWithCertificates(subjCerts, policy, &secTrust);
+  bool result = false;
+  if (ortn) {
+    /* should never happen */
+    goto errOut;
+  }
+
+  result = SecTrustEvaluateWithError(secTrust, nullptr);
+  errOut:
+     if (policy) {
+       CFRelease(policy);
+     }
+  if (secTrust) {
+    CFRelease(secTrust);
+  }
+  if (subjCerts) {
+    CFRelease(subjCerts);
+  }
+  return result;
+}
+
+bool IsCertificateTrustedForPolicy(X509* cert, SecCertificateRef ref) {
+  OSStatus err;
+
+  for (const auto& trust_domain :
+       {kSecTrustSettingsDomainUser, kSecTrustSettingsDomainAdmin}) {
+    CFArrayRef trustSettings;
+    err = SecTrustSettingsCopyTrustSettings(ref, trust_domain, &trustSettings);
+
+    bool isSelfSigned = IsSelfSigned(cert);
+
+    if (err == errSecSuccess && trustSettings != nullptr) {
+      return IsTrustSettingsTrustedForPolicy(trustSettings, isSelfSigned);
+    }
+
+    // An empty trust settings array isn’t the same as no trust settings,
+    // where the trustSettings parameter returns NULL.
+    // No trust-settings array means
+    // “this certificate must be verifiable using a known trusted certificate”.
+    if (trustSettings == nullptr) {
+      return IsCertificateTrustValid(ref);
+    }
+  }
+  return false;
+}
+
+void ReadMacOSKeychainCertificates(
+    std::vector<std::string>* system_root_certificates) {
+  CFTypeRef searchKeys[] = { kSecClass, kSecMatchLimit, kSecReturnRef };
+  CFTypeRef searchValues[] = {
+      kSecClassCertificate, kSecMatchLimitAll, kCFBooleanTrue };
+  CFDictionaryRef search = CFDictionaryCreate(
+      kCFAllocatorDefault, searchKeys, searchValues, 3,
+      &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
+
+  CFArrayRef currAnchors = nullptr;
+  OSStatus ortn = SecItemCopyMatching(
+      search,
+      reinterpret_cast<CFTypeRef *>(&currAnchors));
+
+  if (ortn) {
+      fprintf(stderr, "ERROR: SecItemCopyMatching failed %d\n", ortn);
+  }
+
+  CFIndex count = CFArrayGetCount(currAnchors);
+
+  std::vector<X509*> system_root_certificates_X509;
+  for (int i = 0; i < count ; ++i) {
+    SecCertificateRef certRef = (SecCertificateRef) CFArrayGetValueAtIndex(
+        currAnchors, i);
+
+    CFDataRef derData = SecCertificateCopyData(certRef);
+    if (!derData) {
+      fprintf(stderr, "ERROR: SecCertificateCopyData failed\n");
+      continue;
+    }
+    auto dataBufferPointer = CFDataGetBytePtr(derData);
+
+    X509* cert =
+      d2i_X509(nullptr, &dataBufferPointer, CFDataGetLength(derData));
+    CFRelease(derData);
+    bool isValid = IsCertificateTrustedForPolicy(cert, certRef);
+    if (isValid) {
+      system_root_certificates_X509.emplace_back(cert);
+    }
+  }
+
+  for (size_t i = 0; i < system_root_certificates_X509.size(); i++) {
+    BIOPointer bio(BIO_new(BIO_s_mem()));
```

Comment:
This block can also be mostly simplified using `ncrypto::X509View`

Reply:
Done in https://github.com/nodejs/node/pull/56599/commits/19d30a0e87b6498caf36cf14e9587b7b49901965